hecuba_py/hecuba/IStorage.py

```diff
@@ -90,39 +96,60 @@ def split(self):
-            a partition everytime it's called
+            a partition every time it's called
+            :type tokens_ranges: list[(long,long)]
         """
-        if len(tokens) < min_number_of_tokens:
-            # In this case we have few token and thus we split them
-            tkns_per_partition = min_number_of_tokens / number_of_partitions
-            step_size = ((2 ** 64) - 1) / min_number_of_tokens
-            partition = []
-            for fraction, to in tokens:
-                while fraction < to - step_size:
-                    partition.append((fraction, fraction + step_size))
-                    fraction += step_size
-                    if len(partition) >= tkns_per_partition:
-                        yield partition
-                        partition = []
-                # Adding the last token
-                partition.append((fraction, to))
-            if len(partition) > 0:
-                yield partition
-        else:
-            # This is the case we have more tokens than partitions,.
-            splits = max(len(tokens) / number_of_partitions, 1)
 
-            for i in xrange(0, len(tokens), splits):
-                yield tokens[i:i + splits]
+        tm = config.cluster.metadata.token_map
+        tmap = tm.tokens_to_hosts_by_ks.get(ksp, None)
+        from cassandra.metadata import Murmur3Token
+        tokens_murmur3 = map(lambda a: (Murmur3Token(a[0]), a[1]), tokens_ranges)
+        if not tmap:
+            tm.rebuild_keyspace(ksp, build_if_absent=True)
+            tmap = tm.tokens_to_hosts_by_ks[ksp]
+
+        tokens_per_node = defaultdict(list)
+        for tmumur, t_to in tokens_murmur3:
+            point = bisect_right(tm.ring, tmumur)
+            if point == len(tm.ring):
+                tokens_per_node[tmap[tm.ring[0]][0]].append((tmumur.value, t_to))
+            else:
+                tokens_per_node[tmap[tm.ring[point]][0]].append((tmumur.value, t_to))
+
+        for tokens_in_node in tokens_per_node.values():
+            if len(tokens_in_node) < min_tokens_per_worker:
+                # In this case we have few token and thus we split them
+                step_size = IStorage._max_token // min_tokens_per_worker
+                partition = []
+                for fraction, to in tokens_in_node:
+                    while fraction < to - step_size:
+                        partition.append((fraction, fraction + step_size))
+                        fraction += step_size
+                        if len(partition) >= min_tokens_per_worker:
+                            yield partition
+                            partition = []
+                    # Adding the last token
+                    partition.append((fraction, to))
+                if len(partition) > 0:
+                    yield partition
+            else:
+                # This is the case we have more tokens than partitions,.
+                splits = max(len(tokens_in_node) // min_tokens_per_worker, 1)
```

Comment:
I think we should use a "max_tokens_per_worker", (derived from the variable "NUMBER_OF_BLOCKS") as it was before. I don't see the point of setting a minimum if you can't set a maximum